### PR TITLE
IGNITE-13549 CREATE INDEX/DROP INDEX commands

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ddl/NativeCommandHandler.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        /**
+         * @param schemaSupp Schema supplier.
+         */
+        private SchemaManager(Supplier<SchemaPlus> schemaSupp) {
+            this.schemaSupp = schemaSupp;
+        }
+
+        /** {@inheritDoc} */
+        @Override public GridQueryTypeDescriptor typeDescriptorForTable(String schemaName, String tableName) {
```
Comment (zstan) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ddl/NativeCommandHandler.java:76]: may be enrich code by @Nullable ?

Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ddl/NativeCommandHandler.java:76]: Thanks, added annotation to interface methods and added javadoc.

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -164,6 +164,77 @@ SqlCreate SqlCreateTable(Span s, boolean replace) :
...
+SqlNode IndexedColumn() :
+{
+    final Span s;
+    SqlNode col;
+}
+{
+    col = SimpleIdentifier()
+    (
+        <ASC>
+    |   <DESC> {
```
Comment (korlov42) [inline modules/calcite/src/main/codegen/includes/parserImpls.ftl:176]: do we have any plans to support NULL's ordering?

File: modules/core/src/main/java/org/apache/ignite/internal/sql/SqlCommandProcessor.java [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+import org.apache.ignite.internal.util.future.GridFinishedFuture;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.query.QueryUtils.convert;
+import static org.apache.ignite.internal.processors.query.QueryUtils.isDdlOnSchemaSupported;
+
+/**
+ * Processor responsible for execution of native Ignite commands.
+ */
+public abstract class SqlCommandProcessor {
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/sql/SqlCommandProcessor.java:67]: Why is it abstract?

**Comment by korlov42** [comment]
Looks like we could rework CREATE|DROP TABLE command to use NativeCommandHandler too. WDYT?

File: modules/calcite/src/main/codegen/includes/parserImpls.ftl (hunk 2 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/codegen/includes/parserImpls.ftl:176]: Currently, it's not supported by the Ignite-H2 engine as well and not supported by core indexes. As a first goal, I think we should provide the same functionality as the Ignite-H2 engine, perhaps later NULL's ordering can be added. I can create another ticket if you think this functionality will be helpful.

File: modules/core/src/main/java/org/apache/ignite/internal/sql/SqlCommandProcessor.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/sql/SqlCommandProcessor.java]: Perhaps, it can be used directly. Fixed.

**Comment by alex-plekhanov** [comment]
> Looks like we could rework CREATE|DROP TABLE command to use NativeCommandHandler too. WDYT?

Currently, `CommandProcessor` class supports two types of commands: instances of `SqlCommand` (ignite-core classes) and instances of `GridSqlStatement` (ignite-indexing classes). For some reason (unknown to me) `CREATE/DROP TABLE`, `ALTER TABLE ADD/DROP COLUMN`, and some other commands were implemented as `GridSqlStatement`, such commands can't be used by the calcite engine without additional references to the ignite-indexing module. Commands like `CREATE/DROP INDEX`, `ALTER TABLE LOGGING/NOLOGGING`, 'CREATE/DROP/ALTER USER`, `KILL ...` and some other were implemented as `SqlCommand`, to support such commands in calcite engine we only need to add corresponding grammar and sql-to-cmd converter.
So, we should move the `CREATE/DROP TABLE` command to the core module first (make it instance of 'SqlCommand') to be able to use it in `NativeCommandHandler`. Perhaps such refactoring should be done in the master branch to avoid merge conflicts.

**Comment by korlov42** [comment]
>  Perhaps such refactoring should be done in the master branch to avoid merge conflicts

sounds reasonable

**Review by korlov42 — APPROVED** [review]

**Comment by alex-plekhanov** [comment]
Merged to sql-calcite

**Pull request closed without merging**